### PR TITLE
Keep muon parents in prunedGenParticles collection

### PR DESCRIPTION
**Pull request by kskovpen**
Include additional muon parentage information to be able to correct the muon yields in Muon-enriched QCD MC.

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test
@gpetruc @arizzi please confirm that this addition is ok from miniAOD pov.

**Comment by cvuosalo** [comment]
@kskovpen: I've tried to verify that this PR actually makes an addition to event content, but in all my tests I've seen no increase. What workflow would demonstrate the addition of muon mothers? I tried 38.0, 1313.0, 1320.0, and 1338.0.

**Comment by kskovpen** [comment]
@cvuosalo: Sorry for not mentioning that. The increase would be visible when running on Muon-enriched QCD MC samples. For example, /QCD_Pt-300to470_MuEnrichedPt5_TuneCUETP8M1_13TeV_pythia8/RunIIFall15MiniAODv2-PU25nsData2015v1_76X_mcRun2_asymptotic_v12-v1/MINIAODSIM

File: PhysicsTools/PatAlgos/python/slimming/prunedGenParticles_cfi.py [hunk 1]
```diff
@@ -15,6 +15,7 @@
         "drop   status == 2 && abs(pdgId) == 21",                          # but remove again gluons in the inheritance chain
         "keep abs(pdgId) == 23 || abs(pdgId) == 24 || abs(pdgId) == 25 || abs(pdgId) == 6 || abs(pdgId) == 37 ",   # keep VIP(articles)s
         "keep abs(pdgId) == 310 && abs(eta) < 2.5 && pt > 1 ",                                                     # keep K0
+        "+keep abs(pdgId) == 13 && status == 1", # keep muon parents
```
Comment (slava77) [inline PhysicsTools/PatAlgos/python/slimming/prunedGenParticles_cfi.py]: ``` "++keep abs(pdgId) == 11 || abs(pdgId) == 13 || abs(pdgId) == 15" ``` above is seemingly more inclusive and should have picked up all muons and their parents.
Am I missing some "drop" or is the "++keep abs(pdgId) == 13 " syntax not that inclusive and misses to store muons?

Comment (kskovpen) [inline PhysicsTools/PatAlgos/python/slimming/prunedGenParticles_cfi.py]: @slava77: @gpetruc could confirm, but from what I see is that the inclusive statement you are referring to does not seem to pick up the pion and Kaon parents of muons and one has to specify explicitly the status.

**Comment by slava77** [comment]
Just to repeat earlier call for comments:
 @gpetruc @arizzi please confirm that this addition is ok from miniAOD point of view.

**Comment by gpetruc** [comment]
Sorry I was distracted.

Yes, I'm ok with the change.

AFAIK the reason the first `++keep` doesn't save the parents of the muon is
just that statements are processed in order and so the `drop status == 2`
drops them

Giovanni

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/14384	820d8e6
- code change is as described
- jenkins tests pass and comparisons with baseline show no difference (looks like the change is not happening on the tested workflows based on  https://github.com/cms-sw/cmssw/pull/14384#issuecomment-217585103)
- checked locally in MuEnriched sample, /store/mc/RunIIFall15DR76/QCD_Pt-470to600_MuEnrichedPt5_TuneCUETP8M1_13TeV_pythia8/AODSIM/PU25nsData2015v1_76X_mcRun2_asymptotic_v12-v1/70000/FC95BCDE-799E-E511-BAA9-0025905A60D0.root  in 2K events (running only prunedGenParticles, to not bother trying to make the rest of PAT run on 76X input in 81X) : there is an increase in prunedGenParticlesSize as expected  
```
-----------------------------------------------------------------
   or, B         new, B      delta, B   delta, %   deltaJ, %    branch 
-----------------------------------------------------------------
    291.9 ->       318.7         27      8.8   0.77     recoGenParticlesedmAssociation_prunedGenParticles__PAT.
   3177.7 ->      3402.3        225      6.8   6.48     recoGenParticles_prunedGenParticles__PAT.
-------------------------------------------------------------
     3470 ->        3721        252             7.2     ALL BRANCHES
```

File: PhysicsTools/PatAlgos/python/slimming/prunedGenParticles_cfi.py (hunk 1 above)
Comment (gpetruc) [inline PhysicsTools/PatAlgos/python/slimming/prunedGenParticles_cfi.py]: I think the reason why the ++keep of line 7 is not enough is that the kaons that decay to muons probably have status 2, and so get dropped at line 9.
for a very similar reason we have some more keeps in lines 11 and 12 for electron mothers and for taus.

**Pull request merged**